### PR TITLE
feat: prisma ORM integration

### PR DESCRIPTION
**Pull request by tucker-weed**
[WIP]

Description:

1. Moves db directory to service
2. Adds prisma client to project
3. Adds database functionality tests

TODO:

1. Add more database tests

### CODE REVIEW

File: packages/service/package.json [hunk 1]
```diff
@@ -9,9 +9,11 @@
     "@types/jest": "^26.0.24",
     "babel-jest": "^27.0.6",
     "jest": "^27.0.6",
-    "nodemon": "^2.0.9"
+    "nodemon": "^2.0.9",
+    "prisma": "^2.27.0"
```
Comment (chewnoill) [inline packages/service/package.json:13]: prisma should be a regular dependency I think

Comment (tucker-weed) [reply packages/service/package.json:13]: Their docs say to run `yarn add prisma --dev` - I believe it's because prisma indicated here is just a cli tool

File: db/migrations/20210713211341_add_id_user_document.sql [hunk 2]
```diff
@@ -0,0 +1,10 @@
+-- migrate:up
+CREATE EXTENSION IF NOT EXISTS "uuid-ossp" WITH SCHEMA public;
```
Comment (chewnoill) [inline db/migrations/20210713211341_add_id_user_document.sql:2]: You shouldn't need to create this again, everything from our previous migration will persist.

File: packages/service/package.json [hunk 3]
```diff
@@ -31,6 +31,9 @@
     "y-websocket": "^1.3.15",
     "yjs": "^13.5.10"
   },
+  "prisma": {
+    "schema": "../../db/schema.prisma"
```
Comment (chewnoill) [inline packages/service/package.json:35]: 👍

File: packages/service/package.json (hunk 1 above)
Comment (chewnoill) [reply packages/service/package.json:13]: oh, interesting. 🆒

**Comment by tmsavage** [comment]
Were you able to test the functionality of prisma? Also, would you say this implementation was relatively simple?

**Comment by chewnoill** [comment]
Might be good to add instructions for https://www.prisma.io/studio also

**Comment by tucker-weed** [comment]
> Were you able to test the functionality of prisma? Also, would you say this implementation was relatively simple?

The implementation goes hand in hand with dbmate, which I find to be very simple. Working on tests now.

File: packages/service/src/utils/db.ts [hunk 4]
```diff
@@ -0,0 +1,22 @@
+import { PrismaClient } from "@prisma/client";
+
+const prisma = new PrismaClient();
+
+const db = {
+  create: async (obj, key) => {
+    return await prisma[key].create(obj);
+  },
```
Comment (chewnoill) [inline packages/service/src/utils/db.ts:8]: There are a few problems here.
* this looses a lot of type safety
* it doesn't simply the interface to prisma at all, seems like its just rearranging arguments.

File: packages/service/src/utils/db.test.ts [hunk 5]
```diff
@@ -0,0 +1,31 @@
+import db from "./db";
+
```
Comment (chewnoill) [inline packages/service/src/utils/db.test.ts:2]: there are one to many files named `db.test.ts` here, lets combine them or name them differently.

File: packages/service/src/db.test.ts [hunk 6]
```diff
@@ -1,26 +1,51 @@
...
-	expect(ydoc.getText('codemirror').toJSON()).toBe(
-		ydoc2.getText('codemirror').toJSON()
-	)
-});
\ No newline at end of file
+  // persist this change to the database
+  await db.document.create({
+    data: {
+      value: Buffer.from(state),
+      origin: "test doc",
```
Comment (chewnoill) [inline packages/service/src/db.test.ts:28]: let's flip these around, the value of the document should be the value of the computed CRDT. The encoded CRDT itself should live in the `origin` field.

File: packages/service/src/db.test.ts [hunk 7]
```diff
@@ -1,26 +1,51 @@
...
+      web_rtc_key: "",
+    },
+  });
+
+  // get the change back from the database
+  const stateFromDatabase = (
+    await db.document.findFirst({
+      where: {
+        origin: "test doc",
```
Comment (chewnoill) [inline packages/service/src/db.test.ts:37]: let's select by `id` not by this field.

**Review by chewnoill — COMMENTED** [review]
Nice work @tucker-weed, huge progress here. Couple small things, also want to get the zapatos PR caught up so we can make a better comparison

File: packages/service/src/db.test.ts (hunk 6 above)
Comment (tucker-weed) [reply packages/service/src/db.test.ts:28]: When you say the value of the computed CRDT corresponds to the value field, are you referring to the unencoded CRDT (so just plain document text)?

File: packages/service/package.json [hunk 8]
```diff
@@ -31,6 +31,9 @@
...
+  },
```
Comment (chewnoill) [inline packages/service/package.json:36]: ... does this still make sense?

Comment (tucker-weed) [reply packages/service/package.json:36]: I had that changed, but I think I lost it along the way when structuring commits somehow. Will fix this now

File: readme.md [hunk 9]
```diff
@@ -49,17 +55,27 @@ start storybook
...
+generate prisma schema from dbmate produced schema
+```shell
+yarn orm introspect
 ```
 
-Running migrations
+generate prisma client from prisma schema
 ```shell
-dbmate up
+yarn orm generate
```
Comment (chewnoill) [inline readme.md:78]: let's combine these 3 steps into one. I suggest calling it `codegen`
```json
  "codegen": "yarn migrations up && yarn orm introspect && yarn orm generate"
```

File: readme.md [hunk 10]
```diff
@@ -17,6 +17,12 @@ install packages:
 brew install node docker postgres dbmate
 ```
 
+(optional) install prisma studio
```
Comment (chewnoill) [inline readme.md:20]: do you have to install something here? Why can't you just run `yarn orm studio` or whatever

Comment (tucker-weed) [reply readme.md:20]: Oh - you can do that, just checked. Didn't realize you could open the studio from the command line - I'll update the docs with this change and the change for your comment above

**Comment by chewnoill** [comment]
cherry-picked 5e1f613ae27f967ab1b8b3f2a88453041db46745

File: packages/service/src/db.test.ts [hunk 11]
```diff
@@ -1,26 +0,0 @@
-import * as Y from "yjs";
-
-test("crdts can be persisted", () => {
```
[the remark this answers is not in the document]
Comment (tucker-weed) [reply packages/service/src/db.test.ts:3]: I see what you are saying, I'll make the change

File: packages/service/db/migrations/20210720160437_add_document_updates_queue.sql [hunk 12]
```diff
@@ -0,0 +1,4 @@
+-- migrate:up
+ALTER TABLE document ADD updates_queue bytea[];
```
Comment (chewnoill) [inline packages/service/db/migrations/20210720160437_add_document_updates_queue.sql:2]: I like this idea, but it might make more sense as its own table, instead stuffing it into the document table. Then we can slice/dice updates, and we don't have to worry so much about managing concurrent inserts

File: packages/service/db/schema.prisma [hunk 13]
```diff
@@ -12,6 +12,7 @@ model document {
   value         String
   web_rtc_key   String
   origin        Bytes
+  updates_queue Bytes[]
```
Comment (chewnoill) [inline packages/service/db/schema.prisma:15]: this belongs with the migration

File: packages/service/src/db.test.ts [hunk 14]
```diff
@@ -24,3 +72,85 @@ test("crdts can be persisted", () => {
     ydoc2.getText("codemirror").toJSON()
   );
 });
+
+test("persist equivalent final crdt via different update paths", async () => {
```
Comment (chewnoill) [inline packages/service/src/db.test.ts:76]: Different test, lets put it in its own commit so that it doesn't clutter the diff as much

File: packages/service/src/doc-updates.test.ts [hunk 15]
```diff
@@ -0,0 +1,35 @@
+import * as Y from "yjs";
+
+test("persist equivalent final crdt via different update paths", async () => {
```
Comment (chewnoill) [inline packages/service/src/doc-updates.test.ts:3]: This is a good test for YJS, it's just not testing anything that we wrote.

File: packages/service/src/doc-updates.test.ts [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+  const currentState1 = Y.encodeStateAsUpdate(ydoc1);
+  const currentState2 = Y.encodeStateAsUpdate(ydoc2);
+  ydoc1.destroy();
+  ydoc2.destroy();
+
+  // calculate diffs
+  const stateVector1 = Y.encodeStateVectorFromUpdate(currentState1);
+  const stateVector2 = Y.encodeStateVectorFromUpdate(currentState2);
+  const diff1 = Y.diffUpdate(currentState1, stateVector2);
+  const diff2 = Y.diffUpdate(currentState2, stateVector1);
```
Comment (chewnoill) [inline packages/service/src/doc-updates.test.ts:24]: I'm confused about what this is testing. I think the diffs we care about are the diffs between the original state and the updated state. 

diff1: the diff between state1 origin and state1 updated
diff2: the diff between state2 origin and state2 updated

expect that applying diff2 to state1 will produce the same result as applying diff1 to state2.

Comment (chewnoill) [reply packages/service/src/doc-updates.test.ts:24]: The plan is to store these diffs in the `document_updates_queue` table? Let's write a test for that also.

**Review by chewnoill — DISMISSED** [review]
Looks good, this is ready to merge when/if you guys make the decision

File: packages/service/src/utils/doc-persistence.ts [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+  // find document by id in docs table
+  return db.document.findFirst({
+    where: {
+      id: documentId,
+      document_updates_queue: {
+        every: { document_id: documentId },
+      },
+    },
+    include: {
+      document_updates_queue: true,
```
Comment (chewnoill) [inline packages/service/src/utils/doc-persistence.ts:42]: the join here needs to be on a range I think where `document_updates_queue.created_at > document.latest_update_time`

Comment (tucker-weed) [reply packages/service/src/utils/doc-persistence.ts:42]: Right, good catch
